File: harness/bunnymark/src/main/kotlin/godot/benchmark/bunnymark/v3/Bunny.kt
package godot.benchmark.bunnymark.v3

import godot.RandomNumberGenerator
import godot.Sprite2D
import godot.annotation.GodotScript
import godot.core.Vector2

@GodotScript
class Bunny : Sprite2D() {

	var speed = Vector2()

	private var grav = 500
	private lateinit var screenSize: Vector2
	private val randomNumberGenerator = RandomNumberGenerator()

    override fun _ready() {
        randomNumberGenerator.randomize()
    }

	override fun _process(delta: Double) {
		screenSize = getViewportRect().size
		val pos = position
		val sp = speed

		pos.x += sp.x * delta
		pos.y += sp.y * delta

		sp.y += grav * delta

		if (pos.x > screenSize.x) {
			sp.x *= -1
			pos.x = screenSize.x
		}

		if (pos.x < 0) {
			sp.x *= -1
			pos.x = 0.0
		}

		if (pos.y > screenSize.y) {
			pos.y = screenSize.y
			if (randomNumberGenerator.randf() > 0.5) {
				sp.y = -(randomNumberGenerator.randi() % 1100 + 50).toDouble()
			} else {
				sp.y *= -0.85
			}
		}

		if (pos.y < 0) {
			sp.y = 0.0
			pos.y = 0.0
		}

		position = pos
		speed = sp
	}
}

File: harness/bunnymark/benchmarks/BunnymarkV3/cs/Bunny.cs
using Godot;

// ReSharper disable once CheckNamespace
public class Bunny : Sprite
{
    public Vector2 Speed = Vector2.Zero;

    private readonly int grav = 500;
    private Vector2 screenSize;
    private readonly RandomNumberGenerator randomNumberGenerator = new RandomNumberGenerator();

    public override void _Ready()
    {
        randomNumberGenerator.Randomize();
    }

    public override void _Process(float delta)
    {
        screenSize = GetViewportRect().Size;
        var pos = Position;
        var sp = Speed;

        pos.x += sp.x * delta;
        pos.y += sp.y * delta;

        sp.y += grav * delta;

        if (pos.x > screenSize.x)
        {
            sp.x *= -1f;
            pos.x = screenSize.x;
        }

        if (pos.x < 0)
        {
            sp.x *= -1f;
            pos.x = 0;
        }

        if (pos.y > screenSize.y)
        {
            pos.y = screenSize.y;
            if (randomNumberGenerator.Randf() > 0.5f)
            {
                sp.y = -(randomNumberGenerator.Randi() % 1100f + 50f);
            }
            else
            {
                sp.y *= -0.85f;
            }
        }

        if (pos.y < 0)
        {
            sp.y = 0f;
            pos.y = 0f;
        }

        Position = pos;
        Speed = sp;
    }
}
